fs-index/src/tests.rs

```diff
@@ -0,0 +1,539 @@
+/*!
+# Test Structure for `fs-index` Crate
+
+This test suite is designed to validate the functionality of `fs-index` crate using different hash functions.
+The crate is tested with both cryptographic and non-cryptographic hash functions
+to ensure that the resource index works correctly regardless of the underlying hash function used.
+
+## Modules
+
+- `blake3`: This module contains tests using the `Blake3` hash function.
+- `crc32`: This module contains tests using the `CRC32` hash function.
+- `test_utils`: This module contains common helper macro to generate tests for different hash functions.
+
+*/
+
+mod blake3 {
+    use dev_hash::Blake3;
+
+    use crate::generate_tests_for_hash;
+
+    generate_tests_for_hash!(Blake3);
+}
+
+mod crc32 {
+    use dev_hash::Crc32;
+
+    use crate::generate_tests_for_hash;
+
+    generate_tests_for_hash!(Crc32);
+}
```

Comment:
I think we can improve these tests even further. Core ideas:
- We should not declare separate modules for different hash functions
- We should have more finely-grained control in test cases

----

However, I'm not sure if we can declare lists of hash functions statically in the way that it would be possible to iterate them in test cases. We probably still can use a macro.

- [ ] Define a function, or a macro parameterized by many functions at once, which runs the code section inside, for every function passed to it: `for_each(crc32, blake3, sha256, whatever) { /* test case */ }`

    - If it's a simple for loop through a vector like `for f in vec![crc32, blake3, whatever] { /* test case */ }` that'd be even better, because the code inside the loop could be developed without commenting-out the loop itself. But I'm afraid that it's too difficult to do in Rust
    
- [ ] Define two broad categories of "cryptographic" and "fast" hashes as plain vectors, or as a 2nd-level macros (wrappers), so we can simply add new functions into one of the two categories, and the test cases will work with new functions automatically without necessity to update them manually.

    - This will allow us to describe invariants in generic terms, not specific functions
    ```
    fn test_no_collisions_with_cryptographic_hashing() {
        for_cryptographic_hashing! {
            // setup a folder without duplicates
            // index it
            assert_eq!(index.collisions, 0);
        }
    }
    ```

    - But we are still free to create custom test cases for one or two specific functions, these test cases will not be run excessively for functions that are not involved
    ```
    fn test_no_collisions_with_cryptographic_hashing() {
        for_each_hashing!(take_first_byte, take_last_byte) {
            // setup a folder without duplicates
            // index it
            assert_eq!(index.keys.size(), 8); //synthetic example
        }
    }
    ```

Reply:
Thanks for the detailed review

I've defined a macro `hash_tests` to generate tests for function and hash type pairs, and used it to run tests for the two hash functions currently supported by `dev-hash`.

Adding a new test is straightforward: just add an entry in the macro call with the test function and hash function you want to use.

I've also included instructions on how to navigate the test code and add new tests.

We could use `rstest` to [parameterize tests](https://github.com/la10736/rstest?tab=readme-ov-file#parametrize), but it might not improve code quality much and could add complexity. The current code is well-documented and easy to understand.


Let me know if that's what you had in mind

Reply:
I've seen `rstest` before, good thing but we don't need it right now.

The way how you've implemented works, but curious if you've tried my suggestions above. Does it result in less maintainable codebase?

Reply:
I attempted the first proposed approach. The following dummy code to illustrate:
```rust
/// A macro to take many functions and run them as tests for crc32 and blake3
#[macro_export]
macro_rules! hash_tests_crc32_and_blake3 {
    ($($name:ident,)*) => {
        $(
            for hash_type in &[Crc32, Blake3] {
                #[test]
                fn $name() {
                    $name::<$hash_type>();
                }
            }
        )*
    };
}

hash_tests_crc32_and_blake3!(
    test_store_and_load_index,
    test_store_and_load_index_with_collisions,
    test_build_index_with_file,
    test_build_index_with_empty_file,
    test_build_index_with_directory,
    test_build_index_with_multiple_files,
    test_build_index_with_multiple_directories,
    test_resource_index_update,
    test_add_colliding_files,
    test_num_collisions,
    test_hidden_files,
);
```

This fails to compile with error: 
```shell
macro expansion ignores token `for` and any following
the usage of `hash_tests_crc32_and_blake3!` is likely invalid in item context
```

Attempting to replace it with:
```rust
macro_rules! hash_tests_crc32_and_blake3 {
    ($($name:ident,)*) => {
        $(
            #[test]
            fn $name() {
                $name::<Crc32>();
            }

            #[test]
            fn $name() {
                $name::<Blake3>();
            }
        )*
    };
}
```
or 
```rust
macro_rules! hash_tests_crc32_and_blake3 {
    ($($name:ident,)*) => {
        $(
            #[test]
            fn test_for_crc32() {
                $name::<Crc32>();
            }

            #[test]
            fn test_for_blake3() {
                $name::<Blake3>();
            }
        )*
    };
}
```
fails to compile because multiple test functions have the same name, which Rust doesn't allow. The only way to avoid this error is by separating them into different modules, as in the previous code, but we want to avoid that.

Additionally, we can't dynamically set function names based on the hash type because function names in Rust don't exist at runtime.

While debugging this, I realized we could simply specify the name of the function to the macro, so it's clear. This is the current approach in this PR. I think this gives us fine-grained control over which tests to run and with which hash functions.

Reply:
@tareknaser could you try this approach? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7399c2540956391e1265469e2c2d693e

Ideally, we should be able just iterate through types implementing the trait `ResourceId`. Without generating the actual tests from templated functions.

Reply:
Yes, this compiles fine and does what we need. Let me know what you think of commit d076f34

Reply:
Looks good now, thanks.